### ISSUE TITLE
[HOLD for payment 2024-03-14] [HOLD for payment 2024-02-07] [$500] HIGH: Mark the chat as "read" immediately upon switching to the tab

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#33680]
**Current behaviour:**

1. User is viewing a chat A and opens new tab.
2. User receives messages in chat A and comes back to the tab with chat A being open.
3. Currently the LHN shows as bold, but there is no Unread green line marker.
4. If the user switches to another chat and comes back to the chat A, now chat A behaves as showing a new unread chat. The LHN bold is gone and unread green marker is shown.

**Expected behaviour:**

In step 3, LHN shouldn't be bold, though it should have the green unread marker should be shown to symbolize the new messages received. 

When the user switches away, the green unread marker should be removed. 

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012a54894c108d9212</li>
        <li>Upwork Job ID: 1745132109856296960</li>
        <li>Last Price Increase: 2024-01-10</li>
<li>Automatic offers: </li>
<ul>
<li>abdulrahuman5196 | Reviewer | 28093546</li>
<li>FitseTLT | Contributor | 28093547</li>
</ul></li>
    </ul>
</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Mark the chat as "read" immediately upon switching to the tab

### What is the root cause of that problem?
More of a new feature
### What changes do you think we should make in order to solve the problem?

Currently `currentUnreadMarker` is the state we use to track the new message and we don't reset it when document lose focus so we can reset it when document loses focus when we user switches to other tabs
### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
@quinthar I have some suggestions
1. >Alice is viewing the Bob chat in her NewDot tab, but switches to a different tab to do something else. She gets a message from Bob, and switches back to the tab to view it.
Nope. Now unread marker will only display if you open the report while it has unread messages. If you are already on the report and you receive message while you 're on another tab and switch it back it will not display the marker.
2. >Mark chats as read immediately upon becoming visible in the foreground.

Then what is the reason to display the marker if we remove it immediately, in any other social media apps (e.g. telegram) the marker will exist until you open other chat and I think the current behaviour is correct.

**Comment by bfitzexpensify** [comment]
@abdulrahuman5196 this is one marked as critical, so please prioritise reviewing the proposals here - thanks!

**Comment by quinthar** [comment]
> Nope. Now unread marker will only display if you open the report while it has unread messages. If you are already on the report and you receive message while you 're on another tab and switch it back it will not display the marker.

Ah, good questions!  To clarify, I'm not referring to the "unread marker" (eg, the horizontal line across the chat that designates new from old messages).  I'm referring to it being marked as unread (bold) in the LHN.  I'm sorry for being ambiguous in the original description, does that clarify things?

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Mark the chat as "read" immediately upon switching to the tab

### What is the root cause of that problem?
New Feature:
We already have the logic here 
https://github.com/Expensify/App/blob/88d837c2e88c8c4c22a4c80f93b622d634e67290/src/pages/home/report/ReportActionsList.js#L207-L209

but it only marks the report read if it is `Visibility.isVisible` so once we return from other tab the `useEffect` will not be invoked as it only depends on `report.lastVisibleActionCreated` and `report.reportID` not visibility state
### What changes do you think we should make in order to solve the problem?

We should add visibility change listener and should `Report.readNewestAction` when `Visibility.isVisible`
`Visibility.hasFocus` and the scroll is below the `MSG_VISIBLE_THRESHOLD`

**CAUTION: we should not reset `currentUnreadMarker` as that will remove the new message unread marker which should not be removed (we only want the report to read i.e. remove the boldness in LHN)**

So inside `ReportActionsList` we add this code
```
useEffect(() => {
    const unsubscribeVisibilityListener = Visibility.onVisibilityChange(() => {
        InteractionManager.runAfterInteractions(() => {
            if (!Visibility.isVisible() || !Visibility.hasFocus() || scrollingVerticalOffset.current >= MSG_VISIBLE_THRESHOLD) {
                return;
            }
    
            Report.readNewestAction(report.reportID);
        });
        
    })

    return unsubscribeVisibilityListener;
}, [report.reportID]);
```
or We can also use this code 
https://github.com/Expensify/App/blob/88d837c2e88c8c4c22a4c80f93b622d634e67290/src/pages/home/report/ReportActionsList.js#L207-L209
but we should add visibility state as a dependency here
https://github.com/Expensify/App/blob/88d837c2e88c8c4c22a4c80f93b622d634e67290/src/pages/home/report/ReportActionsList.js#L223C9-L223C57
### What alternative solutions did you explore? (Optional)

**Comment by bfitzexpensify** [comment]
Bump @abdulrahuman5196 - please review these proposals, thank you!

**Comment by FitseTLT** [comment]
> [Proposal updated](https://github.com/Expensify/App/issues/33680#issuecomment-1871189626) with minor change, based on [the requirement update](https://github.com/Expensify/App/issues/33680#issuecomment-1871446463)

>            // put this if we want to clear the unread green marker as well after coming back, if not then ignore it


@abdulrahuman5196 please take a note that this is not a minor change but the main point of the new feature we are implementing. As only making the report unread and not resetting `currentUnreadMarker`  is the critical and main point; and resetting `currentUnreadMarker` will make the unread marker useless as it will make it disappear as soon as the user opens the chat (as the visibility will be focused then) although it should be displayed until the user moves to other report or chat

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
Current Experience: 


https://github.com/Expensify/App/assets/46707890/23d3d4fd-0f9e-4bde-822c-5c3f63aa62da

1) User is viewing a chat A and opens new tab.
2) User receives messages in chat A and comes back to the tab with chat A being open.
3) Currently the LHN shows as bold, but there is no Unread green line marker.
4) If the user switches to another chat and comes back to the chat A, now chat A behaves as showing a new unread chat. Like the LHN bold is gone and unread green marker is shown.


I think there are 2 issues here regarding the LHN and unread marker here

I think in the step 3, we should change the experience like.
**`LHN shouldn't be bold, but green marker should be shown to symbolize the new messages received`**
(which is like a user going to new unread chat)

If we fix step 3, in step 4 the chat would automatically act as already read chat.

Kindly let me if we should change the unread marker as suggested or just have the LHN issue as mentioned in OP.

**Comment by bfitzexpensify** [comment]
@abdulrahuman5196 in step 4, when the user switches away, that would remove the green unread marker?

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 in step 4, when the user switches away, that would remove the green unread marker?

That should be the case IMO, but not happening right now. Atleast in this case if we switch to chat B and come back to chat A after step 3, we are seeing unread marker.
@bfitzexpensify

**Comment by bfitzexpensify** [comment]
Cool, yep, I agree that that should be the case. I think the only other thing to mention is that if the user sends a message on chat A, then we should remove the unread marker, too.

**Comment by FitseTLT** [comment]
Updated proposal according to this [expectation/comment](https://github.com/Expensify/App/issues/33680#issuecomment-1874000153)

**Comment by bfitzexpensify** [comment]
Bump on https://github.com/Expensify/App/issues/33680#issuecomment-1874000153 @abdulrahuman5196. Do you agree with that?

**Comment by abdulrahuman5196** [comment]
> Cool, yep, I agree that that should be the case. I think the only other thing to mention is that if the user sends a message on chat A, then we should remove the unread marker, too. 

Yeah agree. On the user sending a message, I didn't check it, but it should behave the same way as how it behaves now when we send message in a unread chat.

**Comment by abdulrahuman5196** [comment]
@bfitzexpensify  Could you kindly update OP clearly as well as per the discussion here https://github.com/Expensify/App/issues/33680#issuecomment-1873426860

And for contributors kindly update your proposals as per the updated requirements. 
And for the proposals already updated I will check it in my morning alone with if any new.

**Comment by bfitzexpensify** [comment]
OP updated.

**Comment by bfitzexpensify** [comment]
@abdulrahuman5196 were you able to review the existing proposals?

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by quinthar** [comment]
Hey all, what's the ETA?  Who is selecting proposals, and when?

**Comment by abdulrahuman5196** [comment]
Hi, I am reviewing the updated proposals. We have 2 proposals to review. ETA to complete proposal review is today.

**Comment by bfitzexpensify** [comment]
Were you able to review the proposals @abdulrahuman5196?

**Comment by abdulrahuman5196** [comment]
Yes working on it. Will close before EOD.

**Comment by abdulrahuman5196** [comment]
@FitseTLT 's proposal here https://github.com/Expensify/App/issues/33680#issuecomment-1871474470 looks good and works well.
All proposals here relatively similar in core logic. So choosen the first updated proposal which had the functionality for the unread marker and the LHN boldness as mentioned in requirements.

🎀 👀 🎀
C+ Reviewed

**Comment by quinthar** [comment]
Woohoo!  @FitseTLT how long will this take; can you finish it this week?

**Comment by FitseTLT** [comment]
Yep

**Comment by iwiznia** [comment]
🎀 👀 🎀

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 I think [my proposal](https://github.com/Expensify/App/issues/33680#issuecomment-1871189626) is fairly different from the selected one. It makes use of a new hook and much simpler logic, it also doesn't have this whole added logic below which could be buggy.
> 
> I guess jasperhuangg can make the final call here!
> 
> ```
> _.each(sortedReportActions, (reportAction, index) => {
>   if (!shouldDisplayNewMarker(reportAction, index)) {
>       return;
>   }
> 
>   if (!currentUnreadMarker && currentUnreadMarker !== reportAction.reportActionID) {
>       cacheUnreadMarkers.set(report.reportID, reportAction.reportActionID);
>       setCurrentUnreadMarker(reportAction.reportActionID);
>   }
> });
> ```

Hi @tienifr , I would consider code diffs attached with the proposal just as a reference to understand the proposal itself, but not to judge based on its code optimizations which is to be handled at the PR phase(This is also the existing guideline for sometime). So the ask would be to look for the proposal which provided solution for the core issue at hand. If other proposal doesn't have significantly different solution to address the core issue(or significantly higher optimization in some cases), it would be considered duplicate of the original proposal, even though it had some optimizations from original proposal(which could be addressed by the original proposal in PR phase). 

IMO, The current proposal is also choosen based on the same process mentioned above. Yeah, but still the internal engineer can check and revert my decision if he sees your proposal had significantly different solution, which would also be fine by me.

**Comment by FitseTLT** [comment]
> @abdulrahuman5196 I think [my proposal](https://github.com/Expensify/App/issues/33680#issuecomment-1871189626) is fairly different from the selected one. It makes use of a new hook and much simpler logic, it also doesn't have this whole added logic below which could be buggy.
> 
> I guess @jasperhuangg can make the final call here!
> 
> ```
> _.each(sortedReportActions, (reportAction, index) => {
>   if (!shouldDisplayNewMarker(reportAction, index)) {
>       return;
>   }
> 
>   if (!currentUnreadMarker && currentUnreadMarker !== reportAction.reportActionID) {
>       cacheUnreadMarkers.set(report.reportID, reportAction.reportActionID);
>       setCurrentUnreadMarker(reportAction.reportActionID);
>   }
> });
> ```

There is nothing buggy here @tienifr  it is the exact code that is run in the `useEffect` you are suggesting to add screen visibility state as a dependency, I am just explaining the core idea. In fact, we can easily take this code into a separate function and call it  in both the visibility callback and the useEffect without ever needing to create the hook you are suggesting at all (causing a unnecessary rerendering). There is no major difference in the approach in your proposal from mine.

**Comment by hayata-suenaga** [comment]
thank you everyone for proposals 🙇 

this time, as @abdulrahuman5196 suggested, we're going to go with the first uploaded proposal (which is @FitseTLT's)

**Comment by hayata-suenaga** [comment]
@FitseTLT when do you expect you can open a PR?

**Comment by FitseTLT** [comment]
Tomorrow

**Comment by bfitzexpensify** [comment]
I'm going to be mostly ooo until Jan 29, assigning a second BZ team member to keep an eye on this.

- @FitseTLT has been assigned to the issue and is opening a PR today

**Comment by alexpensify** [comment]
Flagging that I'm OOO tomorrow and Monday. I'll check in on Tuesday if there is any action.

**Comment by hayata-suenaga** [comment]
let me know when your PR is ready @FitseTLT 🙇

**Comment by quinthar** [comment]
@FitseTLT given that yesterday didn't happen; what's your new ETA?

**Comment by alexpensify** [comment]
@FitseTLT - please let us know if there is something holding the PR back that we can review to help out. If you are unable to work on this one, we will reassign it. Thanks!

**Comment by FitseTLT** [comment]
Sorry for the delay 🙏 will create one shortly

**Comment by quinthar** [comment]
Can you give a new ETA?  Or should we find someone else to work on this?  Lmk!

**Comment by abdulrahuman5196** [comment]
PR has been created. I will review the PR today.

**Comment by quinthar** [comment]
Awesome!!  When you've reviewed it please update here.   Do you think we can merge today?

**Comment by abdulrahuman5196** [comment]
I reviewed and tested on web. Seems to work good. Will check on other platforms and update. If no issues we would be able to merge today.

**Comment by abdulrahuman5196** [comment]
FYI: I had some comments on the PR. Still discussing on the PR. Will close out as soon as possible.
cc: @hayata-suenaga could you also give a quick look at the discussion.

**Comment by quinthar** [comment]
@hayata-suenaga When can you get to this?  Let's get this out the door!

**Comment by hayata-suenaga** [comment]
I answered in the PR 🙇 @abdulrahuman5196 please complete the reviewer checklist when you have time 🙇

**Comment by abdulrahuman5196** [comment]
Thank you for response. @FitseTLT Could you kindly update the PR as per the latest comments?

**Comment by FitseTLT** [comment]
On it. Give me a min!

**Comment by FitseTLT** [comment]
@abdulrahuman5196 Updated 👍

**Comment by quinthar** [comment]
Great!  With that update, what's the next step, who's doing it, and when?

**Comment by abdulrahuman5196** [comment]
Hi, Regarding the update.
I discovered a bug on the PR on Friday. @FitseTLT Has updated the PR yesterday. 
I am doing the review of the updated changes now and if has fixed the bug. Will update here on any progress.

**Comment by hayata-suenaga** [comment]
I'm also reviewing the PR. I'll give it the final check after @FitseTLT fixes the bug and @abdulrahuman5196 goes through the testing

**Comment by hayata-suenaga** [comment]
also, changing the priority to daily so that we can move through the review process quickly

**Comment by abdulrahuman5196** [comment]
The Bug previously mentioned is still reproducible. @FitseTLT is working on the fix for the same.

https://github.com/Expensify/App/pull/34537#issuecomment-1904688907

**Comment by abdulrahuman5196** [comment]
Had discussions with the contributor offline and the issue is fixed now, And also worked on minor details. I will continue testing in my morning.
Meanwhile I checked on web for usecases and recorded video of the same. Do Let us know if any change is required on any case?


https://github.com/Expensify/App/assets/46707890/cb2e7b12-438a-4910-bff0-9ad7078ed0af

**Comment by quinthar** [comment]
Thanks for keeping the pressure up on this, let's get it out!

**Comment by abdulrahuman5196** [comment]
Starting to review again

**Comment by abdulrahuman5196** [comment]
Status update on the review. We found another issue and fixed it. The current changes looks good. I have tested, added checklist and approved the PR . @hayata-suenaga should be good to start on the internal review.

**Comment by quinthar** [comment]
Awesome, let's find and fix all those bugs!!  @hayata-suenaga what's the ETA for merge?

**Comment by hayata-suenaga** [comment]
review is complete but the performance test is still running. let me merge the PR once it's finished

**Comment by hayata-suenaga** [comment]
The [PR](https://github.com/Expensify/App/pull/34537) was merged

**Comment by alexpensify** [comment]
@abdulrahuman5196 or @hayata-suenaga can we get feedback if this warning was an error or if there is a plan to fix? Thanks!

**Comment by situchan** [comment]
It's deploy blocker (regression from this issue) so should be fixed urgently

**Comment by abdulrahuman5196** [comment]
@FitseTLT Kindly raise a fix PR for the regression bugs caused here https://github.com/Expensify/App/issues/35251 and https://github.com/Expensify/App/issues/35254.  (Both are same issue actually).

**Comment by mountiny** [comment]
Due to two deploy blockers and unclear path for best code solution we have reverted the PR, please address the issues found while ensuring no other regressions or performance degradation would be introduced. Thank you!

**Comment by jeremy-croff** [comment]
Testing this on my branch: https://github.com/Expensify/App/pull/35004 it will fix the LHN being bold. But it doesn't address the unread marker. I believe 2 seperate logics.

**Comment by abdulrahuman5196** [comment]
Not overdue. @FitseTLT is working on the regression fix

**Comment by bfitzexpensify** [comment]
Sweet. Looks like you're up on reviewing the fix @abdulrahuman5196

**Comment by alexpensify** [comment]
Whoops, I missed that @bfitzexpensify is back online: 

https://github.com/Expensify/App/issues/33680#issuecomment-1887652810

I'm stepping away here and Ben will deal with the BZ next steps when that time comes.

**Comment by bfitzexpensify** [comment]
Mind taking a look at the [PR](https://github.com/Expensify/App/pull/35948) @abdulrahuman5196?

**Comment by abdulrahuman5196** [comment]
Hi @bfitzexpensify Sorry for the delay. Will check on it today.

**Comment by bfitzexpensify** [comment]
@abdulrahuman5196 what's the latest with this issue?

**Comment by abdulrahuman5196** [comment]
@bfitzexpensify  new changes have been merged yesterday

**Comment by hayata-suenaga** [comment]
yep [the pr](https://github.com/Expensify/App/pull/37021) was merged

**Comment by bfitzexpensify** [comment]
Payments complete, please complete the BZ checklist when you get a moment @abdulrahuman5196 - thank you!

**Comment by abdulrahuman5196** [comment]
>The PR that introduced the bug has been identified. Link to the PR:
The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

Not a regression. New implementation.

>Determine if we should create a regression test for this bug.

Yes.

>If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.

1) Login with User B and open a chat with User A and change to another tab on web
2) Now Send a message as User A to User B
3) Wait until the message arrives at User B and now as User B switch back to the tab you were chatting in (1)
4) Check that the LHN is not bold
5) verify that unread marker indicates on the new message you sent in (2)

@bfitzexpensify

**Comment by abdulrahuman5196** [comment]
I think we can close this @bfitzexpensify
